### ISSUE TITLE
Allow 0 and negative indices for GridPosition

### ISSUE THREAD

**Issue by knuesel** [MakieOrg/Makie.jl#1534]
Currently, the second line in

```julia
scatter(f[1,1], ...)
Label(f[0,1], ...)
```

will create a label at the top and "move" the existing elements downwards. I find it a bit odd:

* it gives a simple way to add elements at the top or left (without reference to the current grid size) ~~but there's no equivalent for the bottom or right~~ (Edit: actually we case already use `end+1` for that)
* most importantly: it "invalidates" the previous indices, which can turn the code into a spaghetti mess. Consider

   ```julia
   scatter(f[1,1], ...)
   ...
   Label(f[1,0], ...)
   ...
   scatter(f[3,2], ...)
   ...
   Label(f[0, 2])
   ```

   The positioning has become imperative instead of declarative. It's hard to follow which command corresponds to which position in the end result: one has to mentally execute the whole code.

Instead, shouldn't `Label(f[0,1], ...)` simply create a label at position (0,1)? And to add a line before that we would then write `Label(f[-1,1], ...)`, etc. These zero or negative indices would then be valid grid positions, and the meaning of existing indices would not change.

We could still have a simple way to append things on either side of the grid, using `begin` and `end`. For example `Label(f[begin-1, end+1], ...)` would add the label in a new row at the top and a new column at the right.

**Comment by jkrumbiegel** [comment]
It's a good idea, it is actually annoying that the labels shift, I agree. Do you think OffsetArrays would fix the `begin` `end` issue? That's the main reason why I didn't do this, I have not worked with OffsetArrays, yet, though.

**Comment by jkrumbiegel** [comment]
Although actually, as I'm not storing things in a matrix, this is all my own implementation anyway, I just need to store the offset indices and adjust the rest of the column / row logic accordingly

**Comment by knuesel** [comment]
Ah then OffsetArrays is not really relevant indeed (except as an example of non-standard array indices).

For the `begin`/`end` issue, actually `end` (and also `end+1`, etc.) is already working thanks to this `lastindex` implementation:

https://github.com/JuliaPlots/Makie.jl/blob/9a4821bcaf1cf0f09841e659dd2a6eb81afe6527/src/figures.jl#L112

To get `begin` to work, you'd just need to implement `firstindex` similarly. Or you could remove the `lastindex` implementation and implement `axes(::Figure, i)` instead, since `firstindex` and `lastindex` both have fallbacks that use `axes`.

**Comment by jkrumbiegel** [comment]
This should do it https://github.com/jkrumbiegel/GridLayoutBase.jl/pull/29 once it's merged and tagged, although it will be a breaking GridLayoutBase version, so we'd need 0.17 for Makie.

**Comment by knuesel** [comment]
Awesome, it seems to work well from my limited testing with GridLayoutBase.

**Comment by knuesel** [comment]
This has been implemented a while ago.
